### PR TITLE
Restructure of taxa worksheet handling

### LINKED ISSUE #78: Revert banned fields in GBIF Taxa sheet
The use of extra fields in the GBIF taxa sheet was permitted in the past but is now prevented. This was actually used a few times and is supportable, so we should re-enable it. This is at least partly to allow re-validation of legacy datasets

### PR DESCRIPTION
**Pull request by davidorme**
Turned out to be a lot more involved!

* GBIFTaxa now allows any extra columns
* So does NCBITaxa but...
    * Taxon handling extended to allow any NCBI taxon rank as a header, with the exception of clade and no rank, which have no consistent placement in the ranking and can be duplicated in the hierarchy.
    * The `NCBIValidator` is extended to contain an ordered list of those ranks, which is used to order tuples of provided rank information, along with exhaustive lists of valid kingdom and superkingdom names.
    * Those lists are used to handle the of commonplace Kingdom to Superkingdom label promotion issue at the `NCBITaxa` worksheet level: tackling individual rows.  The `NCBIValidator` now does not need to attempt to handle this condition and so validates standard NCBI ranks.
    * Adds ability to flag a leaf taxon as new, which gets it added to the taxon index with a validated parent.
    * Restructure of `NCBIValidator` functionality to break out reused code into some private methods/static methods.
    * A couple of functions now raise errors rather than returning None on error, so are handled with `try` blocks.
    
The flow is basically unchanged, so hopefully it isn't too hard to follow the fairly big changes!

**Note** - I haven't updated the docs to match yet, in case it is all wrong.

Closes #78

### CODE REVIEW

File: test/test_ncbi_taxa.py [hunk 1]
```diff
@@ -16,304 +16,651 @@
...
+            id="Backbone leaf",
         ),
-        (  # Genus name not in binomial
-            dict(higher_nm="Canis", lower_nm="Vulpes vulpes", tri=False),
+        pytest.param(
+            562,
+            does_not_raise(),
+            ("Escherichia coli 1-110-08_S1_C1", "strain", 1444049, 562),
+            9,
+            id="Non backbone leaf",
```
Comment (jacobcook1995) [inline test/test_ncbi_taxa.py:405]: I might be misunderstanding the test but I don't get this one, I don't understand how searching `tax_id=562` will return `("Escherichia coli 1-110-08_S1_C1", "strain", 1444049, 562)` as the first element of the hierarchy?

File: test/test_ncbi_taxa.py [hunk 2]
```diff
@@ -386,284 +695,316 @@ def test_id_lookup_errors(fixture_ncbi_validator, test_input, expected_exception
...
+        #     dict(
+        #         nnme="Opistho",
+        #         taxon_hier=[("superkingdom", "Eukaryota"), ("clade", "Opisthokonta")],
+        #     ),
+        #     ("clade", "Opisthokonta", 33154, 2759),
+        #     does_not_raise(),
+        #     (None, None),
+        #     ((INFO, "Match found for Opistho"),),
+        #     id="Clade anchored by superkingdom",
+        # ),
```
Comment (jacobcook1995) [inline test/test_ncbi_taxa.py:774]: Guess this one should either be commented back in or deleted

File: test/test_ncbi_taxa.py [hunk 3]
```diff
@@ -386,284 +695,316 @@ def test_id_lookup_errors(fixture_ncbi_validator, test_input, expected_exception
...
-            ((ERROR, "More than one possible parent taxa for A Morus found"),),
-        ),
-        # Eukaryote Morus
-        (
-            dict(nnme="E Morus", taxah={"superkingdom": "Eukaryota", "genus": "Morus"}),
-            (
                 (
                     ERROR,
-                    "Parent taxa for E Morus refers to multiple possible child taxa",
+                    "Taxonomy mismatch for Cytophaga at rank genus: expecting "
```
Comment (jacobcook1995) [inline test/test_ncbi_taxa.py:855]: I do worry about this one. Basically if someone is uploading old data and a species has been moved to a new genus, do they have to update the entire hierarchy? For kingdom -> family I guess not, but with genus you run into the problem that the species binomial gets checked against the genus name

File: test/test_ncbi_taxa.py [hunk 4]
```diff
@@ -683,775 +1024,430 @@ def test_taxa_search_errors(fixture_ncbi_validator, test_input, expected_excepti
...
         ),
-        # Same idea for a non-backbone case
-        (
-            ["Streptophytina", {"phylum": "Streptophytina"}, None],
+        # # Taxon hierarchy in wrong order
+        # (
+        #     ["E coli", {"species": "Escherichia coli", "genus": "Escherichia"}, None],
+        #     does_not_raise(),
+        #     ((ERROR, "Taxon hierarchy not in correct order"),),
+        # ),
```
Comment (jacobcook1995) [inline test/test_ncbi_taxa.py:1289]: Guess either comment this back in or remove it

File: test/test_ncbi_taxa.py [hunk 5]
```diff
@@ -1537,87 +1525,268 @@ def test_validate_index_higher_taxa(
...
+            (
+                (INFO, "Loading NCBITaxa worksheet"),
+                (INFO, "Reading NCBI taxa data"),
+                (INFO, "3 NCBI rank fields found:"),
+                (INFO, "Validating row 1:"),
+                (ERROR, "Rank family has non-string or empty string value: 123"),
+                (INFO, "NCBITaxa contains 1 errors"),
+            ),
+            id="non string taxon",
+        ),
```
Comment (jacobcook1995) [inline test/test_ncbi_taxa.py:1756]: Guess a useful test here is whether if you give the full outdated hierarchy for `Cytophaga marina` (i.e. including Cytophaga as the genus) whether it passes or fails

**Review by jacobcook1995 — COMMENTED** [review]
Think moving away from the old structure (which was aiming to as far as possible mirror the `GBIF` validation) is a sensible move and worth the effort it took. My comments are basically a load of minor points and then a few queries.

File: test/test_ncbi_taxa.py (hunk 3 above)
Comment (jacobcook1995) [reply test/test_ncbi_taxa.py:855]: Though maybe this is handled by using timestamped database versions

File: safedata_validator/taxa.py [hunk 6]
```diff
@@ -251,77 +223,73 @@ def __post_init__(self):
...
-                )
-                is False
-            ):
-                raise ValueError("Taxa tuples not all in [string integer integer] form")
+        if (
+            isinstance(self.ncbi_id, float) and not self.ncbi_id.is_integer()
+        ) or not isinstance(
+            self.ncbi_id, int
+        ):  # Catch non int or float case
+            raise TypeError("NCBI ID is not an integer")
```
[the remark this answers is not in the document]
Comment (davidorme) [reply safedata_validator/taxa.py:231]: One thing I meant to say on the PR - some comments in the code though - is that the NCBITaxon/Validator code is very defensive about the inputs - I guess we'd generally expect these to be fed cleaned inputs. I don't see any real problem with that though - it isn't really much of a bottleneck.

File: safedata_validator/taxa.py [hunk 7]
```diff
@@ -552,404 +549,372 @@ def id_lookup(self, nnme: str, ncbi_id: int) -> NCBITaxon:
...
-                mtaxon = self.id_lookup(nnme, tID)
-            # Then check whether multiple taxonomic levels have been provided
-            else:
-                # If not raise an error
-                LOGGER.error(
-                    f"Taxa {nnme} cannot be found using only one "
-                    f"taxonomic level, more should be provided"
+        # Now ensure the taxon hierarchy is in root to leaf order and then get the leaf
+        # and look for that name and rank combination in the database.
+        taxon_hier.sort(key=lambda x: self.ncbi_ranks_root_to_leaf.index(x[0]))
```
[the remark this answers is not in the document]
Comment (davidorme) [reply safedata_validator/taxa.py:803]: Yes it does - the NCBI taxonomy db creation routine contains a sorting algorithm that will fail if this isn't definitive. Users do not need to follow order in the workbook as the fields get sorted.

File: safedata_validator/taxa.py [hunk 8]
```diff
@@ -1057,31 +1022,24 @@ def load(self, worksheet: worksheet) -> None:
...
-            LOGGER.error(
-                "Unexpected (or misspelled) headers found:",
-                extra={"join": unexpected_headers},
-            )
-            FORMATTER.pop()
-            return
+        # Now check for extra fields and report them to the user
+        extra_fields = set(headers).difference(tx_fields)
+        if extra_fields:
+            LOGGER.info("Additional fields provided: ", extra={"join": extra_fields})
```
[the remark this answers is not in the document]
Comment (davidorme) [reply safedata_validator/taxa.py:1042]: I think so - that list is intended to allow people to spot `gneus` typo issues and the like, so having a complete list does no harm, I think.

File: safedata_validator/taxa.py [hunk 9]
```diff
@@ -1671,90 +1628,50 @@ def load(self, worksheet: worksheet) -> None:
...
             FORMATTER.pop()
             return
 
-        # Possible fields in this case are the two core fields + comments + all
-        # taxonomic ranks defined in NCBI
-        all_fields = set(
-            list(core_fields) + BACKBONE_RANKS_EX + EXTRA_NCBI_RANKS + ["comments"]
-        )
+        # The set of possible data fields is name, new and the NCBI ranks
+        data_fields = ["name", "new"] + self.validator.ncbi_ranks_root_to_leaf
```
[the remark this answers is not in the document]
Comment (davidorme) [reply safedata_validator/taxa.py:1638]: Yes - for now we do nothing with NCBI ID, so that seems appropriate?

File: safedata_validator/taxa.py [hunk 10]
```diff
@@ -1671,90 +1628,50 @@ def load(self, worksheet: worksheet) -> None:
...
 
-        # Field cleaning
-        core_fields = {"name", "ncbi id"}
-        missing_core = core_fields.difference(headers)
-
-        if missing_core:
-            # core names are not found so can't continue
-            LOGGER.error("Missing core fields: ", extra={"join": missing_core})
+        # Only the name field is indispensible
+        if "name" not in headers:
```
[the remark this answers is not in the document]
Comment (davidorme) [reply safedata_validator/taxa.py:1632]: Good catch.

File: safedata_validator/taxa.py [hunk 11]
```diff
@@ -1854,207 +1855,106 @@ def validate_and_add_taxon(self, ncbi_taxon_input: tuple) -> None:
...
+                self.taxon_index.append(
+                    (
+                        m_name,
+                        hr_taxon.ncbi_id,
+                        hr_taxon.parent_ncbi_id,
+                        hr_taxon.non_canon_name,
+                        hr_taxon.rank,
+                        hr_taxon.non_canon_name_class,
+                    )
+                )
```
[the remark this answers is not in the document]
Comment (davidorme) [reply safedata_validator/taxa.py:1955]: Yes. Same is done for GBIF.

File: test/test_ncbi_taxa.py (hunk 2 above)
Comment (davidorme) [reply test/test_ncbi_taxa.py:774]: Yeah - I don't see how this can currently occur since we don't allow `clade` fields in the worksheet at present.

File: safedata_validator/taxa.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jacobcook1995) [reply safedata_validator/taxa.py:1955]: Makes sense, just thought I should check

File: safedata_validator/taxa.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jacobcook1995) [reply safedata_validator/taxa.py:1042]: Fair enough

File: test/test_ncbi_taxa.py (hunk 3 above)
Comment (davidorme) [reply test/test_ncbi_taxa.py:855]: Yeah - the timestamping, I think. I don't see how we'd handle it otherwise.

File: safedata_validator/taxa.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jacobcook1995) [reply safedata_validator/taxa.py:1638]: Oh right I hadn't clocked that, then yeah I agree it should be reported

File: test/test_ncbi_taxa.py (hunk 2 above)
Comment (jacobcook1995) [reply test/test_ncbi_taxa.py:774]: Yeah I don't see why anyone would want to report a clade, so think test can go

File: test/test_ncbi_taxa.py (hunk 1 above)
Comment (davidorme) [reply test/test_ncbi_taxa.py:405]: Test was borked and not actually testing that the `leaf` value matched. Which of course it didn't.

File: test/test_ncbi_taxa.py (hunk 3 above)
Comment (jacobcook1995) [reply test/test_ncbi_taxa.py:855]: Yeah I suppose it's better here than in the GBIF case as is less a matter of opinion and more a matter of when they did their analysis, which is something time stamping allows us to account for

File: test/test_ncbi_taxa.py (hunk 5 above)
Comment (jacobcook1995) [reply test/test_ncbi_taxa.py:1756]: Or probably not if we are using time stamping

File: docs/data_providers/data_format/ncbi_taxa.md [hunk 12]
```diff
@@ -50,64 +50,79 @@ taxonomic usage!) but the NCBI backbone has extremely good taxonomic coverage.
...
+  You are only required to provide the taxonomic name for the specific rank that you are
+  trying to match and can leave other fields empty. However it is probably more useful
+  to provide a more complete taxonomy! If you do provide higher taxonomic information
+  then it must be congruent with the hierarchy for the focal taxon. For example,
+  specifying Family Anatidae is sufficient to identify waterfowl, but providing Order
+  Carnivora (rather than Order Anseriformes) would result in an error.
+
+  Names can be provided in plain text, or alternatively in a commonly used notation,
+  where the rank is indicated by a lower case first letter and the name follows after
+  two underscores (e.g. k__Bacteria for Kingdom Bacteria). Notation of this type should
```
Comment (jacobcook1995) [inline docs/data_providers/data_format/ncbi_taxa.md:101]: I guess k__Bacteria should probably be in ``

File: docs/data_providers/data_format/ncbi_taxa.md [hunk 13]
```diff
@@ -50,64 +50,79 @@ taxonomic usage!) but the NCBI backbone has extremely good taxonomic coverage.
...
+  be placed in the correct rank columns, and validation is carried out to check that the
+  rank implied by the notation matches the column rank.
+
+  Two special cases are that NCBI outputs typically separate out the components of
+  binomial and trinomial names: for example, they might return `g_Escherichia` and
```
Comment (jacobcook1995) [inline docs/data_providers/data_format/ncbi_taxa.md:106]: g__Escherichia

File: docs/data_providers/data_format/ncbi_taxa.md [hunk 14]
```diff
@@ -50,64 +50,79 @@ taxonomic usage!) but the NCBI backbone has extremely good taxonomic coverage.
...
+  your genus field contains `Escherichia` and your species field contains `Escherichia
+  coli` then this will also be accepted, as long as the parts are compatible.
+  
+  !!! Note
+
+    Missing rank entries are completely fine, e.g. leaving out phylum information for
+    some taxa but providing it for others. However, sufficient information must be
+    provided to unambiguously identify each taxon.
+
+* **Comments and other field**: These fields are obviously optional. If you
```
Comment (jacobcook1995) [inline docs/data_providers/data_format/ncbi_taxa.md:120]: "fields"?

**Review by jacobcook1995 — APPROVED** [review]
LGTM! Just made a note of a few possible typos
